src/sentry/users/web/accounts.py

```diff
@@ -374,3 +390,84 @@ def confirm_email(request: HttpRequest, user_id: int, hash: str) -> HttpResponse
         )
     messages.add_message(request, level, msg)
     return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+
+
+@set_referrer_policy("strict-origin-when-cross-origin")
+@login_required
+@control_silo_function
+def confirm_signed_email(
+    request: HttpRequest, signed_data: str
+) -> HttpResponseRedirect | HttpResponse:
+    EMAIL_CONFIRMATION_SALT = options.get("user-settings.signed-url-confirmation-emails-salt")
+
+    use_signed_urls = options.get("user-settings.signed-url-confirmation-emails")
+    if not use_signed_urls:
+        msg = ERR_CONFIRMING_EMAIL
+        level = messages.ERROR
+        messages.add_message(request, level, msg)
+        return HttpResponseRedirect(reverse("sentry-account-settings-emails"))
+
+    msg = _("Thanks for confirming your email")
+    level = messages.SUCCESS
+
+    try:
+        data = unsign(
+            signed_data, salt=EMAIL_CONFIRMATION_SALT, max_age=172800
+        )  # max age is 2 days in seconds
+
+        # is the currently logged in user the one that
+        # wants to add the email to their account
+        if request.user.id != int(data["user_id"]):
+            raise InvalidRequest
```

Comment:
Should this raise a different HTTP status code like not enough permission or something?

Reply:
Never mind I just saw that all of these cause a redirect so it doesn't matter.